ttnn/cpp/ttnn/operations/core.hpp

```diff
@@ -152,58 +155,309 @@ inline ttnn::Tensor squeeze_from_4D(const ttnn::Tensor& tensor, const int rank)
+        for (int i = 0; i < output_tensors.size(); i++) {
+            if (output_tensors[i].has_value()) {
+                output_tensors[i].value().tensor_attributes->update_main_thread_ref_count(
+                    workers.at(0), output_tensor_ref_count[i]);
+            }
+        }
     }
 
-    const auto original_shape = tensor.get_shape();
-    const auto tensor_4D = unsqueeze_to_4D(tensor);
-
-    if (memory_config.is_sharded()) {
-        // to_sharded path
-        if (tensor_4D.is_sharded()) {
-            // reshard
-            const auto input_memory_config = ttnn::get_memory_config(tensor_4D);
-            const auto input_shard_spec = input_memory_config.value().shard_spec.value();
-            const auto output_shard_spec = memory_config.shard_spec.value();
-            if (tensor_4D.get_layout() == ttnn::TILE_LAYOUT ||
-                input_shard_spec.shape[1] == output_shard_spec.shape[1]) {
-                if (dtype.has_value()) {
-                    throw runtime_error("dtype cannot be specified when converting sharded tensor to sharded tensor");
+    template <class ConcreteOperation>
+    static ProgramOutputTensors<ConcreteOperation> convert_before_run(const OptionalTensors& tensors) {
+        if constexpr (std::is_same_v<OptionalTensors, ProgramOutputTensors<ConcreteOperation>>) {
+            return tensors;
+        } else {
+            ProgramOutputTensors<ConcreteOperation> output_tensors;
+            for (const auto& tensor : tensors) {
+                if (tensor.has_value()) {
+                    output_tensors.push_back(tensor.value());
                 }
-                return reshape(tt::tt_metal::reshard(tensor_4D, memory_config), original_shape);
-            } else {
-                // for row-major tensors where shard-spec[1] is different for input shard and output shard
-                return reshape(
-                    tt::tt_metal::interleaved_to_sharded(
-                        tt::tt_metal::sharded_to_interleaved(tensor_4D, ttnn::DRAM_MEMORY_CONFIG, dtype),
-                        memory_config,
-                        dtype),
-                    original_shape);
             }
+            return output_tensors;
+        }
+    }
+
+    template <class ConcreteOperation>
+    static ProgramOutputTensors<ConcreteOperation> convert_before_run(const Tensors& tensors) {
+        if constexpr (std::is_same_v<Tensors, ProgramOutputTensors<ConcreteOperation>>) {
+            return tensors;
+        } else {
+            ProgramOutputTensors<ConcreteOperation> output_tensors;
+            for (const auto& tensor : tensors) {
+                output_tensors.push_back(tensor);
+            }
+            return output_tensors;
+        }
+    }
+
+    template <class ConcreteOperation>
+    static OptionalTensors convert_after_run(const ProgramOutputTensors<ConcreteOperation>& tensors) {
+        if constexpr (std::is_same_v<OptionalTensors, ProgramOutputTensors<ConcreteOperation>>) {
```

Comment:
Could you please explain the purpose of these conversion calls here. I'm not entirely sure about what's going on here.

Reply:
We plan to transition from using std::vector<Tensor> for both output and input tensors to std::vector<std::optional<Tensor>> for both output and input tensors.   The intent is to make the input and output for each operation symetric.  The purpose of the conversion calls is to simply make this a bit easier to transition.  Essentially, the launch_op itself will have to be written or possibly depreciated.